File: android/src/main/java/com/datadog/reactnative/DdRum.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.reactnative

import com.datadog.android.bridge.DdBridge
import com.datadog.android.bridge.DdRum as SDKDdRum
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap

/**
 * The entry point to use Datadog's RUM feature.
 */
class DdRum(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private val nativeInstance: SDKDdRum = DdBridge.getDdRum(reactContext)

    override fun getName(): String = "DdRum"

    /**
     * Start tracking a RUM View.
     * @param key The view unique key identifier.
     * @param name The view name.
     * @param timestamp The timestamp when the view started.
     * @param context The additional context to send.
     */
    @ReactMethod
    fun startView(key: String, name: String, timestamp: Double, context: ReadableMap, promise: Promise) {
        nativeInstance.startView(key, name, timestamp.toLong(), context.toHashMap())
        promise.resolve(null)
    }

    /**
     * Stop tracking a RUM View.
     * @param key The view unique key identifier.
     * @param timestamp The timestamp when the view stopped.
     * @param context The additional context to send.
     */
    @ReactMethod
    fun stopView(key: String, timestamp: Double, context: ReadableMap, promise: Promise) {
        nativeInstance.stopView(key, timestamp.toLong(), context.toHashMap())
        promise.resolve(null)
    }

    /**
     * Start tracking a RUM Action.
     * @param type The action type (tap, scroll, swipe, click, custom).
     * @param name The action name.
     * @param timestamp The timestamp when the action started.
     * @param context The additional context to send.
     */
    @ReactMethod
    fun startAction(type: String, name: String, timestamp: Double, context: ReadableMap, promise: Promise) {
        nativeInstance.startAction(type, name, timestamp.toLong(), context.toHashMap())
        promise.resolve(null)
    }

    /**
     * Stop tracking the ongoing RUM Action.
     * @param timestamp The timestamp when the action stopped.
     * @param context The additional context to send.
     */
    @ReactMethod
    fun stopAction(timestamp: Double, context: ReadableMap, promise: Promise) {
        nativeInstance.stopAction(timestamp.toLong(), context.toHashMap())
        promise.resolve(null)
    }

    /**
     * Add a RUM Action.
     * @param type The action type (tap, scroll, swipe, click, custom).
     * @param name The action name.
     * @param timestamp The timestamp when the action occured.
     * @param context The additional context to send.
     */
    @ReactMethod
    fun addAction(type: String, name: String, timestamp: Double, context: ReadableMap, promise: Promise) {
        nativeInstance.addAction(type, name, timestamp.toLong(), context.toHashMap())
        promise.resolve(null)
    }

    /**
     * Start tracking a RUM Resource.
     * @param key The resource unique key identifier.
     * @param method The resource method (GET, POST, …).
     * @param url The resource url.
     * @param timestamp The timestamp when the resource started.
     * @param context The additional context to send.
     */
    @ReactMethod
    fun startResource(key: String, method: String, url: String, timestamp: Double, context: ReadableMap, promise: Promise) {
        nativeInstance.startResource(key, method, url, timestamp.toLong(), context.toHashMap())
        promise.resolve(null)
    }

    /**
     * Stop tracking a RUM Resource.
     * @param key The resource unique key identifier.
     * @param statusCode The resource status code.
     * @param kind The resource's kind (xhr, document, image, css, font, …).
     * @param timestamp The timestamp when the resource stopped.
     * @param context The additional context to send.
     */
    @ReactMethod
    fun stopResource(key: String, statusCode: Double, kind: String, timestamp: Double, context: ReadableMap, promise: Promise) {
        nativeInstance.stopResource(key, statusCode.toLong(), kind, timestamp.toLong(), context.toHashMap())
        promise.resolve(null)
    }

    /**
     * Add a RUM Error.
     * @param message The error message.
     * @param source The error source (network, source, console, logger, …).
     * @param stacktrace The error stacktrace.
     * @param timestamp The timestamp when the error occured.
     * @param context The additional context to send.
     */
    @ReactMethod
    fun addError(message: String, source: String, stacktrace: String, timestamp: Double, context: ReadableMap, promise: Promise) {
        nativeInstance.addError(message, source, stacktrace, timestamp.toLong(), context.toHashMap())
        promise.resolve(null)
    }

}

File: packages/core/ios/Sources/DdRum.m
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(DdRum, NSObject)

RCT_EXTERN_METHOD(startView:(NSString)key
                 withName:(NSString)name
                 withContext:(NSDictionary)context
                 withTimestampms:(NSInteger)timestampMs
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(stopView:(NSString)key
                 withContext:(NSDictionary)context
                 withTimestampms:(NSInteger)timestampMs
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(startAction:(NSString)type
                 withName:(NSString)name
                 withContext:(NSDictionary)context
                 withTimestampms:(NSInteger)timestampMs
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(stopAction:(NSString)type
                 withName:(NSString)name
                 withContext:(NSDictionary)context
                 withTimestampms:(NSInteger)timestampMs
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(addAction:(NSString)type
                 withName:(NSString)name
                 withContext:(NSDictionary)context
                 withTimestampms:(NSInteger)timestampMs
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(startResource:(NSString)key
                 withMethod:(NSString)method
                 withUrl:(NSString)url
                 withContext:(NSDictionary)context
                 withTimestampms:(NSInteger)timestampMs
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(stopResource:(NSString)key
                 withStatuscode:(NSInteger)statusCode
                 withKind:(NSString)kind
                 withSize:(NSInteger)size
                 withContext:(NSDictionary)context
                 withTimestampms:(NSInteger)timestampMs
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(addError:(NSString)message
                 withSource:(NSString)source
                 withStacktrace:(NSString)stacktrace
                 withContext:(NSDictionary)context
                 withTimestampms:(NSInteger)timestampMs
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(addTiming:(NSString)name
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(stopSession:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

@end
